iroh-bytes/src/store/redb/validate.rs

```diff
@@ -0,0 +1,362 @@
+//! Validation of the store's contents.
+//!
+//! This performs a full consistency check. Eventually it will also validate
+//! file content again, but that part is not yet implemented.
+use std::collections::BTreeSet;
+
+use redb::ReadableTable;
+
+use crate::store::{ValidateLevel, ValidateProgress};
+
+use super::{
+    raw_outboard_size, tables::ReadableTables, ActorResult, ActorState, DataLocation, EntryState,
+    Hash, OutboardLocation,
+};
+
+impl ActorState {
+    pub(super) fn validate(
+        &mut self,
+        tables: &impl ReadableTables,
+        progress: tokio::sync::mpsc::Sender<ValidateProgress>,
+    ) -> ActorResult<()> {
+        let blobs = tables.blobs();
+        let inline_data = tables.inline_data();
+        let inline_outboard = tables.inline_outboard();
+        let tags = tables.tags();
+        macro_rules! send {
+        ($level:expr, $entry:expr, $($arg:tt)*) => {
+            if let Err(_) = progress.blocking_send(ValidateProgress::ConsistencyCheckUpdate { message: format!($($arg)*), level: $level, entry: $entry }) {
+                return Ok(());
+            }
+        };
+    }
+        macro_rules! trace {
+        ($($arg:tt)*) => {
+            send!(ValidateLevel::Trace, None, $($arg)*)
+        };
+    }
+        macro_rules! info {
+        ($($arg:tt)*) => {
+            send!(ValidateLevel::Info, None, $($arg)*)
+        };
+    }
+        macro_rules! warn {
+        ($($arg:tt)*) => {
+            send!(ValidateLevel::Warn, None, $($arg)*)
+        };
+    }
+        macro_rules! entry_warn {
+        ($hash:expr, $($arg:tt)*) => {
+            send!(ValidateLevel::Warn, Some($hash), $($arg)*)
+        };
+    }
+        macro_rules! entry_info {
+        ($hash:expr, $($arg:tt)*) => {
+            send!(ValidateLevel::Info, Some($hash), $($arg)*)
+        };
+    }
+        macro_rules! error {
+        ($($arg:tt)*) => {
+            send!(ValidateLevel::Error, None, $($arg)*)
+        };
+    }
+        macro_rules! entry_error {
+        ($hash:expr, $($arg:tt)*) => {
+            send!(ValidateLevel::Error, Some($hash), $($arg)*)
+        };
+    }
+        // first, dump the entire data content at trace level
```

Comment:
shouldn't this be gated on the log level, currently this is executed no matter the log level

Reply:
The log level is not part of the rpc msg. Filtering happens on the client side. I think for something like fsck this is fine. Compared to the other stuff this does it won't matter in terms of perf. And validate is special anyway since it locks the db for an extended period.